### ISSUE TITLE
From source build error OSX

### ISSUE THREAD

**Issue by bzcheeseman** [pytorch/pytorch#2322]
I'm getting a not found error during my build from source, here is the stack trace:
```
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingHostAllocator.cpp:58:42: error: a space is required between consecutive right angle
      brackets (use '> >')
  std::deque<std::pair<cudaEvent_t, void*>> cuda_events;
                                         ^~
                                         > >
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingHostAllocator.cpp:74:5: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    auto it = available.lower_bound(search_key);
    ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:99:43: error: a space is required between consecutive right angle
      brackets (use '> >')
  std::deque<std::pair<cudaEvent_t, Block*>> cuda_events;
                                          ^~
                                          > >
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingHostAllocator.cpp:93:19: error: expected expression
    blocks.insert({*ptr, Block(size, *ptr, true)});
                  ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingHostAllocator.cpp:111:5: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    auto it = blocks.find(ptr);
    ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingHostAllocator.cpp:139:5: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    auto it = blocks.find(ptr);
    ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingHostAllocator.cpp:163:7: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
      auto& e = cuda_events.front();
      ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingHostAllocator.cpp:192:10: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    for (auto it = cuda_events.begin(); it != cuda_events.end(); ++it) {
         ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingHostAllocator.cpp:208:10: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    for (auto it = blocks.begin(); it != blocks.end();) {
         ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingHostAllocator.cpp:228:10: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    for (auto it = streams.begin(); it != streams.end(); ++it) {
         ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingHostAllocator.cpp:229:7: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
      auto& stream = *it;
      ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingHostAllocator.cpp:242:19: error: no member named 'emplace_back' in
      'std::__1::deque<std::__1::pair<CUevent_st *, void *>, std::__1::allocator<std::__1::pair<CUevent_st *, void *> > >'
      cuda_events.emplace_back(event, block.ptr);
      ~~~~~~~~~~~ ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:125:5: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    auto& free_blocks = small ? large_blocks : small_blocks;
    ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:130:5: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    auto it = free_blocks.lower_bound(&search_key);
    ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:174:5: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    auto it = allocated_blocks.find(ptr);
    ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:232:5: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    auto it = blocks.lower_bound(&search_key);
    ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:270:5: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    auto& free_blocks = small ? large_blocks : small_blocks;
    ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:361:9: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
        auto cur = it;
        ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:373:5: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    auto it = allocated_blocks.find(ptr);
    ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:390:10: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
    for (auto it = streams.begin(); it != streams.end(); ++it) {
         ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:391:7: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
      auto& stream = *it;
      ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:404:19: error: no member named 'emplace_back' in
      'std::__1::deque<std::__1::pair<CUevent_st *, (anonymous namespace)::Block *>, std::__1::allocator<std::__1::pair<CUevent_st *,
      (anonymous namespace)::Block *> > >'
      cuda_events.emplace_back(event, block);
      ~~~~~~~~~~~ ^
/Users/Aman/code/pytorch/pytorch/torch/lib/THC/THCCachingAllocator.cpp:419:7: warning: 'auto' type specifier is a C++11 extension
      [-Wc++11-extensions]
      auto& e = cuda_events.front();
      ^
8 warnings and 3 errors generated.
make[2]: *** [CMakeFiles/THC.dir/THCCachingHostAllocator.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
10 warnings and 2 errors generated.
make[2]: *** [CMakeFiles/THC.dir/THCCachingAllocator.cpp.o] Error 1
make[1]: *** [CMakeFiles/THC.dir/all] Error 2
make: *** [all] Error 2
```

I cloned the pytorch repo master branch. This error occurred with both `clang-3.7` (installed via homebrew) as well as Apple clang
```
Apple LLVM version 8.1.0 (clang-802.0.42)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

My NVCC version is
```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2016 NVIDIA Corporation
Built on Tue_Jan_10_13:22:46_CST_2017
Cuda compilation tools, release 8.0, V8.0.61
```

The exact commands I ran were these:
`MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang CXX=clang++ python setup.py install`
and
`MACOSX_DEPLOYMENT_TARGET=10.9 CC=clang-3.7 CXX=clang++-3.7 python setup.py install`

It makes it through libTH, libTHNN, libTHS, and crashes on libTHC

**Comment by killeent** [comment]
For some reason I think your clang++ compiler is not compiling with `-std=c++11`. Can you move the line specifying the standard in `build_all.sh` from https://github.com/pytorch/pytorch/blob/master/torch/lib/build_all.sh#L140 to line 125?

**Comment by bzcheeseman** [comment]
It definitely got further this time - this is the error I get now:
```
clang-3.7 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/Users/Aman/anaconda3/include -arch x86_64 -I/Users/Aman/anaconda3/include -arch x86_64 -I/Users/Aman/code/pytorch/pytorch -I/Users/Aman/code/pytorch/pytorch/torch/csrc -I/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include -I/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/TH -I/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/THPP -I/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/THNN -I/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/ATen -I/Users/Aman/anaconda3/lib/python3.6/site-packages/numpy/core/include -I/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/THD -I/usr/local/cuda/include -I/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/THCUNN -I/usr/local/cuda/include -I/Users/Aman/anaconda3/include/python3.6m -c torch/csrc/autograd/functions/utils.cpp -o build/temp.macosx-10.7-x86_64-3.6/torch/csrc/autograd/functions/utils.o -D_THP_CORE -std=c++11 -Wno-write-strings -fno-strict-aliasing -DWITH_NUMPY -DWITH_DISTRIBUTED -DWITH_CUDA -DCUDA_LIB_PATH=/usr/local/cuda/lib -DWITH_CUDNN
torch/csrc/autograd/functions/convolution.cpp:185:20: error: no viable conversion from 'std::vector<long>' to 'IntList'
      (aka 'ArrayRef<int64_t>')
    output.resize_(output_size(input, weight));
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/ATen/ArrayRef.h:34:9: note: candidate constructor (the implicit copy constructor)
      not viable: no known conversion from 'std::vector<long>' to 'const at::ArrayRef<long long> &' for 1st argument
  class ArrayRef {
        ^
/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/ATen/ArrayRef.h:34:9: note: candidate constructor (the implicit move constructor)
      not viable: no known conversion from 'std::vector<long>' to 'at::ArrayRef<long long> &&' for 1st argument
  class ArrayRef {
        ^
/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/ATen/ArrayRef.h:57:18: note: candidate constructor not viable: no known
      conversion from 'std::vector<long>' to 'const long long &' for 1st argument
    /*implicit*/ ArrayRef(const T &OneElt)
                 ^
/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/ATen/ArrayRef.h:83:18: note: candidate constructor not viable: no known
      conversion from 'std::vector<long>' to 'const std::initializer_list<long long> &' for 1st argument
    /*implicit*/ ArrayRef(const std::initializer_list<T> &Vec)
                 ^
/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/ATen/ArrayRef.h:70:18: note: candidate template ignored: could not match
      'long long' against 'long'
    /*implicit*/ ArrayRef(const std::vector<T, A> &Vec)
                 ^
/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/ATen/ArrayRef.h:75:28: note: candidate template ignored: could not match 'array'
      against 'vector'
    /*implicit*/ constexpr ArrayRef(const std::array<T, N> &Arr)
                           ^
/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/ATen/ArrayRef.h:80:28: note: candidate template ignored: could not match
      'const long long [N]' against 'std::vector<long>'
    /*implicit*/ constexpr ArrayRef(const T (&Arr)[N]) : Data(Arr), Length(N) {}
                           ^
/Users/Aman/code/pytorch/pytorch/torch/lib/tmp_install/include/ATen/TensorMethods.h:13:41: note: passing argument to parameter 'size' here
inline Tensor & Tensor::resize_(IntList size) {
```

**Comment by killeent** [comment]
Ok, this is a KP as with #2299 and is fixed with #2319. I will open a PR to move the flags higher up to address your first issue.
